// element-plus/lib/utils/props.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.componentSize = exports.mutable = exports.keyOf = exports.definePropType = exports.buildProps = exports.buildProp = exports.propKey = void 0;
const shared_1 = require("@vue/shared");
const mapValues_1 = __importDefault(require("lodash/mapValues"));
const error_1 = require("./error");
const wrapperKey = Symbol();
exports.propKey = Symbol();
function buildProp(option) {
    var _a;
    if (!(0, shared_1.isObject)(option) || !!option[exports.propKey])
        return option;
    const { values, required, default: defaultValue, type, validator } = option;
    const _validator = values || validator
        ? (val) => {
            let valid = false;
            let allowedValues = [];
            if (values) {
                allowedValues = [...values, defaultValue];
                valid || (valid = allowedValues.includes(val));
            }
            if (validator)
                valid || (valid = validator(val));
            if (!valid && allowedValues.length > 0) {
                (0, error_1.debugWarn)(`Vue warn`, `Invalid prop: Expected one of (${allowedValues.join(', ')}), got value ${val}`);
            }
            return valid;
        }
        : undefined;
    return {
        type: ((_a = type) === null || _a === void 0 ? void 0 : _a[wrapperKey]) || type,
        required: !!required,
        default: defaultValue,
        validator: _validator,
        [exports.propKey]: true,
    };
}
exports.buildProp = buildProp;
const buildProps = (options) => (0, mapValues_1.default)(options, (option) => buildProp(option));
exports.buildProps = buildProps;
const definePropType = (val) => ({ [wrapperKey]: val });
exports.definePropType = definePropType;
const keyOf = (arr) => Object.keys(arr);
exports.keyOf = keyOf;
const mutable = (val) => val;
exports.mutable = mutable;
exports.componentSize = ['large', 'medium', 'small', 'mini'];
