// element-plus/packages/hooks/dist/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useTimeout = exports.useTeleport = exports.usePreventGlobal = exports.useThrottleRender = exports.useFocus = exports.useMigrating = exports.useModal = exports.useRestoreActive = exports.useLockScreen = exports.useEvents = exports.useAttrs = void 0;
var use_attrs_1 = require("./use-attrs");
Object.defineProperty(exports, "useAttrs", { enumerable: true, get: function () { return __importDefault(use_attrs_1).default; } });
var use_events_1 = require("./use-events");
Object.defineProperty(exports, "useEvents", { enumerable: true, get: function () { return __importDefault(use_events_1).default; } });
var use_lockscreen_1 = require("./use-lockscreen");
Object.defineProperty(exports, "useLockScreen", { enumerable: true, get: function () { return __importDefault(use_lockscreen_1).default; } });
var use_restore_active_1 = require("./use-restore-active");
Object.defineProperty(exports, "useRestoreActive", { enumerable: true, get: function () { return __importDefault(use_restore_active_1).default; } });
var use_modal_1 = require("./use-modal");
Object.defineProperty(exports, "useModal", { enumerable: true, get: function () { return __importDefault(use_modal_1).default; } });
var use_migrating_1 = require("./use-migrating");
Object.defineProperty(exports, "useMigrating", { enumerable: true, get: function () { return __importDefault(use_migrating_1).default; } });
var use_focus_1 = require("./use-focus");
Object.defineProperty(exports, "useFocus", { enumerable: true, get: function () { return __importDefault(use_focus_1).default; } });
var use_throttle_render_1 = require("./use-throttle-render");
Object.defineProperty(exports, "useThrottleRender", { enumerable: true, get: function () { return __importDefault(use_throttle_render_1).default; } });
var use_prevent_global_1 = require("./use-prevent-global");
Object.defineProperty(exports, "usePreventGlobal", { enumerable: true, get: function () { return __importDefault(use_prevent_global_1).default; } });
var use_teleport_1 = require("./use-teleport");
Object.defineProperty(exports, "useTeleport", { enumerable: true, get: function () { return __importDefault(use_teleport_1).default; } });
var use_timeout_1 = require("./use-timeout");
Object.defineProperty(exports, "useTimeout", { enumerable: true, get: function () { return __importDefault(use_timeout_1).default; } });
__exportStar(require("./use-model-toggle"), exports);
__exportStar(require("./use-popper"), exports);
__exportStar(require("./use-css-var"), exports);
__exportStar(require("./use-locale"), exports);
__exportStar(require("./use-form-item"), exports);
__exportStar(require("./use-same-target"), exports);
